src/Authentication/JWTAuthenticator.php

```diff
@@ -25,201 +34,252 @@
 
 class JWTAuthenticator extends MemberAuthenticator
 {
+    use Injectable;
     use Configurable;
+    use RequiresConfig;
+    use MemberTokenGenerator;
 
     /**
-     * @var Sha256|RsaSha256
+     * Set to true to allow anonymous JWT tokens (no member record / email / password)
+     *
+     * @config
+     * @var bool
+     */
+    private static $anonymous_allowed = false;
+
+    /**
+     * @config
+     * @var int
      */
-    private $signer;
+    private static $nbf_time = 0;
 
     /**
-     * @var string|Key;
+     * Expires after 1 hour
+     *
+     * @config
+     * @var int
      */
-    private $privateKey;
+    private static $nbf_expiration = 3600;
 
     /**
-     * @var string|Key;
+     * Token can be refreshed within 7 days
+     *
+     * @config
+     * @var int
      */
-    private $publicKey;
+    private static $nbf_refresh_expiration = 604800;
 
     /**
-     * JWTAuthenticator constructor.
-     * @throws JWTException
+     * @return Signer
      */
-    public function __construct()
+    protected function getSigner(): Signer
     {
-        $key = Environment::getEnv('JWT_SIGNER_KEY');
-        if (empty($key)) {
-            throw new JWTException('No key defined!', 1);
+        $signerKey = $this->getEnv('JWT_SIGNER_KEY');
+        if (PathResolver::resolve($signerKey)) {
+            return new Rsa\Sha256();
+        } else {
+            return new Hmac\Sha256();
         }
-        $publicKeyLocation = Environment::getEnv('JWT_PUBLIC_KEY');
-        if (file_exists($key) && !file_exists($publicKeyLocation)) {
-            throw new JWTException('No public key found!', 1);
+    }
+
+    /**
+     * Get private key
+     *
+     * @return Key
+     */
+    protected function getPrivateKey(): Key
+    {
+        $signerKey = $this->getEnv('JWT_SIGNER_KEY');
+        $signerPath = PathResolver::resolve($signerKey);
+        if ($signerPath) {
+            $password = $this->getEnv('JWT_KEY_PASSWORD', null);
+            return new Key('file://' . $signerPath, $password);
         }
+        return new Key($signerKey);
     }
 
     /**
-     * Setup the keys this has to be done on the spot for if the signer changes between validation cycles
+     * Get public key
+     *
+     * @return Key
+     * @throws LogicException
      */
-    private function setKeys()
+    private function getPublicKey(): Key
```

Comment:
I think it would be good to unify this method and simply use `getKey($type)` where type defines if it's public or private. It would reduce code duplication IMO

Reply:
Yeah what did I do here. Normally I'm the one who complains about this...

Although public and private keys are different; I can probably still clean this code up.

Reply:
Result is nice; I'll push it up soon.

I did add a getKeyType() to consolidate the if / else logic. getPublicKey() / getPrivateKey() just call the consolidated method.